### DISCUSSION TITLE
Combining Cyfrin and Chainlink projects?

### DISCUSSION THREAD

**Discussion by whatTHEh3ll** [Cyfrin/foundry-full-course-cu#1121] [Q&A]
I'm trying to create an on chain nft with Foundry. I'm basing my project on the Cyfrin NFT repo.

Except that I'm not doing the Moodnft. I decided to go with a Smart contract from another Patrick Collins project. Here [https://github.com/PatrickAlphaC/all-on-chain-generated-nft](https://github.com/PatrickAlphaC/all-on-chain-generated-nft)

I would like to create a basic on chain svg with foundry. The contract seems to compile correctly. However when I execute a test I get an error on the '**testTokenURIIsCorrect()**' function. Reason assertion violated.

Here's the test [https://gitlab.com/b2896/foundry-nft/-/blob/main/test/SvgNftTwoTest.t.sol](https://gitlab.com/b2896/foundry-nft/-/blob/main/test/SvgNftTwoTest.t.sol)

My project repo [https://gitlab.com/b2896/foundry-nft/](https://gitlab.com/b2896/foundry-nft/)

Any ideas on writing a successful unit test and/or deploy script?

Thank you

**Comment by 0xusmanf** [comment]
Your test is failing because the two URI's you are expecting to be the same in your test aren't actually the same because when you call the `svgNftTwo.create(JSON_URI);` function it does not save the URI as it is but instead it applies 2 methods (`svgToImageURI`,`formatTokenURI`) on `JSON_URI` before saving it.

Hence you have to apply these 2 methods to the `JSON_URI` in your test file, and then you can expect the assert condition to pass like in the function below.
```solidity
    function testTokenURIIsCorrect() public {
        vm.prank(USER);
        svgNftTwo.create(JSON_URI);
        string memory uri = svgNftTwo.svgToImageURI(JSON_URI);
        string memory final_uri = svgNftTwo.formatTokenURI(uri);

        assert(keccak256(abi.encodePacked(svgNftTwo.tokenURI(0))) == keccak256(abi.encodePacked(final_uri)));
    }
```

**Reply by whatTHEh3ll** [reply]
That worked! Thank you!
